src/term/transform/float_combinators.rs

```diff
@@ -1,277 +1,151 @@
+        scope.push(nam);
+      }
+      current = bod;
     }
 
-    if depth != 0 && detach.can_detach() && term_info.has_no_free_vars() {
-      term_info.detach_term(self);
+    match current {
+      Term::Var { nam } if scope.contains(&nam) => true,
+      Term::Ref { .. } => true,
+      term => term.is_safe(book, seen),
     }
-
-    term_info.merge_scope(parent_scope);
-
-    detach
   }
 
-  fn float_app(&mut self, depth: usize, term_info: &mut TermInfo) -> Detach {
-    let Term::App { fun, arg, .. } = self else { unreachable!() };
-
-    let parent_scope = term_info.replace_scope(BTreeSet::new());
-
-    let fun_detach = fun.go_float(depth + 1, term_info);
-    let fun_scope = term_info.replace_scope(BTreeSet::new());
-
-    let arg_detach = arg.go_float(depth + 1, term_info);
-    let arg_scope = term_info.replace_scope(parent_scope);
-
-    let detach = match fun_detach {
-      // If the fun scope is not empty, we know the recursive ref is not in a active position,
-      // Because if it was an application, it would be already extracted
-      //
-      // e.g.: {recursive_ref some_var}
-      Detach::Recursive if !fun_scope.is_empty() => arg_detach,
-
-      Detach::Recursive if arg_scope.is_empty() && arg_detach.can_detach() => {
-        term_info.detach_term(self);
-        Detach::Combinator
-      }
-
-      // If the only term that is possible to detach is just a Term::Ref,
-      // there is no benefit to it, so that case is skipped
-      Detach::Recursive if !matches!(fun, box Term::Ref { .. }) => {
-        term_info.detach_term(fun);
-        arg_detach
-      }
-
-      _ => fun_detach & arg_detach,
-    };
-
-    term_info.merge_scope(fun_scope);
-    term_info.merge_scope(arg_scope);
-
-    detach
+  /// A term is a supercombinator if it is a lambda abstraction without free variables.
+  pub fn is_supercombinator(&self) -> bool {
+    matches!(self, Term::Lam { .. } if self.free_vars().is_empty())
   }
 
-  fn float_mat(&mut self, depth: usize, term_info: &mut TermInfo) -> Detach {
-    let Term::Mat { args, rules } = self else { unreachable!() };
-
-    let mut detach = Detach::Combinator;
-    for arg in args {
-      detach = detach & arg.go_float(depth + 1, term_info);
-    }
-    let parent_scope = term_info.replace_scope(BTreeSet::new());
-
-    for rule in rules.iter_mut() {
-      for pat in &rule.pats {
-        debug_assert!(pat.is_native_num_match());
-      }
-
-      let arm_detach = match rule.body.go_float(depth + 1, term_info) {
-        // If the recursive ref reached here, it is not in a active position
-        Detach::Recursive => Detach::Combinator,
-        detach => detach,
-      };
-
-      detach = detach & arm_detach;
-    }
+  pub fn has_unscoped_diff(&self) -> bool {
+    let (declared, used) = self.unscoped_vars();
+    declared.difference(&used).count() != 0
+  }
 
-    term_info.merge_scope(parent_scope);
-    detach
+  fn is_combinator(&self) -> bool {
```

Comment:
I think it's fine to consider as combinators other non-lambda terms, like matches, tups, sups, etc. Particularly, it would be useful to extract dups and lets, which can become redexes and so cause loop.
Concretely, changing `is_combinator` to just not having free variables or not having free variables and not having unmatched unscoped vars/binds.